sw/device/lib/testing/test_framework/ottf_console.c

```diff
@@ -0,0 +1,167 @@
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/base/status.h"
+#include "sw/device/lib/dif/dif_rv_plic.h"
+#include "sw/device/lib/dif/dif_uart.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_isrs.h"
+#include "sw/device/lib/testing/test_framework/ottf_test_config.h"
+
+// TODO: make this toplevel agnostic.
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+/**
+ * OTTF console configuration parameters.
+ */
+enum {
+  /**
+   * Flow control parameters.
+   */
+  kFlowControlLowWatermark = 4,   // bytes
+  kFlowControlHighWatermark = 8,  // bytes
+  kFlowControlRxWatermark = kDifUartWatermarkByte8,
+  /**
+   * HART PLIC Target.
+   */
+  kPlicTarget = kTopEarlgreyPlicTargetIbex0,
+};
+
+// Potential DIF handles for OTTF console communication.
+static dif_uart_t ottf_console_uart;
+
+// The `flow_control_state` and `flow_control_irqs` variables are shared between
+// the interrupt service handler and user code.
+static volatile ottf_console_flow_control_t flow_control_state;
+static volatile uint32_t flow_control_irqs;
+
+void *ottf_console_get() {
+  switch (kOttfTestConfig.console.type) {
+    default:
+      return &ottf_console_uart;
+  }
+}
+
+void ottf_console_init(void) {
+  // Initialize/Configure the console device.
+  uintptr_t base_addr = kOttfTestConfig.console.ip_base_addr;
+  switch (kOttfTestConfig.console.type) {
+    case (kOttfConsoleUart):
+      // Set a default for the console base address if the base address is not
+      // configured. The default is to use UART0.
+      if (base_addr == 0) {
+        base_addr = TOP_EARLGREY_UART0_BASE_ADDR;
+      }
+      CHECK_DIF_OK(
+          dif_uart_init(mmio_region_from_addr(base_addr), &ottf_console_uart));
+      CHECK_DIF_OK(dif_uart_configure(&ottf_console_uart,
+                                      (dif_uart_config_t){
+                                          .baudrate = kUartBaudrate,
+                                          .clk_freq_hz = kClockFreqPeripheralHz,
+                                          .parity_enable = kDifToggleDisabled,
+                                          .parity = kDifUartParityEven,
+                                          .tx_enable = kDifToggleEnabled,
+                                          .rx_enable = kDifToggleEnabled,
+                                      }));
+      base_uart_stdout(&ottf_console_uart);
+      break;
+    default:
+      CHECK(false, "unsupported OTTF console interface.");
+      break;
+  }
+
+  // Initialize/Configure console flow control (if requested).
+  if (kOttfTestConfig.enable_uart_flow_control) {
+    ottf_console_flow_control_enable();
+  }
+}
+
+void ottf_console_flow_control_enable(void) {
```

Comment:
Ah yes, I saw that. Since we hardcode the use of UART0 when the UART is selected as the console device this should be ok for now, no? But I can certainly create a follow up PR that makes the exact console device configurable, so we can use different UARTs, etc.

Reply:
actually let met add to this PR, its super quick fix.

Reply:
fixed.